Docker image build failing

### Problem

The Autonity 0.5.0 docker image build is failing in dockerhub.

logs (https://hub.docker.com/repository/registry-1.docker.io/clearmatics/autonity/builds/4a403b58-e02c-47b3-8cc0-019991db1142?ref=login) :
```
Cloning into '.'...
Warning: Permanently added the RSA host key for IP address '140.82.114.4' to the list of known hosts.
Submodule 'tests' (https://github.com/ethereum/tests) registered for path 'tests/testdata'
Cloning into 'tests/testdata'...
Submodule path 'tests/testdata': checked out '5841af6da472fb3f19810354cf9a30afd8e72b5f'
Switched to a new branch 'master'
Pulling cache layers for index.docker.io/clearmatics/autonity:latest...
Done!
KernelVersion: 4.4.0-1060-aws
Components: [{u'Version': u'19.03.8', u'Name': u'Engine', u'Details': {u'KernelVersion': u'4.4.0-1060-aws', u'Os': u'linux', u'BuildTime': u'2020-03-11T01:24:30.000000000+00:00', u'ApiVersion': u'1.40', u'MinAPIVersion': u'1.12', u'GitCommit': u'afacb8b7f0', u'Arch': u'amd64', u'Experimental': u'false', u'GoVersion': u'go1.12.17'}}, {u'Version': u'1.2.13', u'Name': u'containerd', u'Details': {u'GitCommit': u'7ad184331fa3e55e52b890ea95e65ba581ae3429'}}, {u'Version': u'1.0.0-rc10', u'Name': u'runc', u'Details': {u'GitCommit': u'dc9208a3303feef5b3839f4323d9beb36df0a9dd'}}, {u'Version': u'0.18.0', u'Name': u'docker-init', u'Details': {u'GitCommit': u'fec3683'}}]
Arch: amd64
BuildTime: 2020-03-11T01:24:30.000000000+00:00
ApiVersion: 1.40
Platform: {u'Name': u'Docker Engine - Community'}
Version: 19.03.8
MinAPIVersion: 1.12
GitCommit: afacb8b7f0
Os: linux
GoVersion: go1.12.17
Starting build of index.docker.io/clearmatics/autonity:latest...
Step 1/9 : FROM golang:1.14-alpine as builder
---> 3289bf11c284
Step 2/9 : RUN apk add --no-cache make gcc musl-dev linux-headers libc-dev git
---> Running in 552ed13beebf
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/community/x86_64/APKINDEX.tar.gz
(1/21) Upgrading musl (1.1.24-r8 -> 1.1.24-r9)
(2/21) Installing libgcc (9.3.0-r2)
(3/21) Installing libstdc++ (9.3.0-r2)
(4/21) Installing binutils (2.34-r1)
(5/21) Installing gmp (6.2.0-r0)
(6/21) Installing isl (0.18-r0)
(7/21) Installing libgomp (9.3.0-r2)
(8/21) Installing libatomic (9.3.0-r2)
(9/21) Installing libgphobos (9.3.0-r2)
(10/21) Installing mpfr4 (4.0.2-r4)
(11/21) Installing mpc1 (1.1.0-r1)
(12/21) Installing gcc (9.3.0-r2)
(13/21) Installing nghttp2-libs (1.41.0-r0)
(14/21) Installing libcurl (7.69.1-r0)
(15/21) Installing expat (2.2.9-r1)
(16/21) Installing pcre2 (10.35-r0)
(17/21) Installing git (2.26.2-r0)
(18/21) Installing musl-dev (1.1.24-r9)
(19/21) Installing libc-dev (0.7.2-r3)
(20/21) Installing linux-headers (5.4.5-r1)
(21/21) Installing make (4.3-r0)
Executing busybox-1.31.1-r16.trigger
OK: 163 MiB in 35 packages
Removing intermediate container 552ed13beebf
---> 96cbbe4d0931
Step 3/9 : ADD . /autonity
---> 167e2c8c5250
Step 4/9 : RUN cd /autonity && make autonity
---> Running in 0b32f500cf30
mkdir -p ./common/acdefault/generated
[91mmake: sudo: No such file or directory
make: *** [Makefile:47: common/acdefault/generated/bytecode.go] Error 127
[0m
sudo docker run --rm -v /autonity/./contracts/autonity/contract/contracts:/contracts -v /autonity/./common/acdefault/generated:/output ethereum/solc:0.6.4 --overwrite --abi --bin -o /output /contracts/Autonity.sol
Removing intermediate container 0b32f500cf30
The command '/bin/sh -c cd /autonity && make autonity' returned a non-zero code: 2
```

### Acceptance criteria

The image can be built on dockerhub.

### Question

Could we have this build integrated into our own C.I.?

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -369,3 +369,30 @@ jobs:
 
       - name: test contract
         run: make test-contracts
+
+  # Builds the docker image. We don't intend to use the built image as part of
+  # a workflow, but we want to know the image can be built.
+  docker-image-build:
+    # Don't try to build the docker image unless autonity itself builds.
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - name: Cache code
+        uses: actions/cache@v1
+        env:
+          cache-name: cache-autonity-code
+        with:
+          path: /home/runner/work/autonity/autonity/
+          key: ${{ runner.os }}-aut-${{ env.cache-name }}-${{ github.sha }}
+
+      - uses: actions/checkout@v2
+        if: steps.cache-autonity-code.outputs.cache-hit != 'true'
+        with:
+          clean: 'false'
+
+      - name: Embed Autonity contract
+        if:  steps.cache-autonity-code.outputs.cache-hit != 'true'
+        run: make embed-autonity-contract
+
+      - name: Build
+        run: make build-docker-image

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -4,6 +4,18 @@ jobs:
   bootstrap-checkout:
     runs-on: ubuntu-latest
     steps:
+      - name: Set up Go
+        uses: actions/setup-go@v1
+        with:
+          go-version: 1.13
+        id: go
+
+      - name: Set GOPATH
+        run: |
+          echo "##[set-env name=GOPATH;]$(dirname $GITHUB_WORKSPACE)"
+          echo "##[add-path]$(dirname $GITHUB_WORKSPACE)/bin"
+        shell: bash
+
       - name: Cache code
         uses: actions/cache@v1
         id: cache-autonity-code

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -383,28 +383,17 @@ jobs:
         run: make test-contracts
 
   # Builds the docker image. We don't intend to use the built image as part of
-  # a workflow, but we want to know the image can be built.
+  # a workflow, but we want to know the image can be built. We don't want to
+  # use any cache here since we want a clean checkout before any code
+  # generation has taken place.
   docker-image-build:
-    # Don't try to build the docker image unless autonity itself builds.
     runs-on: ubuntu-latest
+    # Don't try to build the docker image unless autonity itself builds.
     needs: build
     steps:
-      - name: Cache code
-        uses: actions/cache@v1
-        env:
-          cache-name: cache-autonity-code
-        with:
-          path: /home/runner/work/autonity/autonity/
-          key: ${{ runner.os }}-aut-${{ env.cache-name }}-${{ github.sha }}
-
       - uses: actions/checkout@v2
-        if: steps.cache-autonity-code.outputs.cache-hit != 'true'
         with:
           clean: 'false'
-
-      - name: Embed Autonity contract
-        if:  steps.cache-autonity-code.outputs.cache-hit != 'true'
-        run: make embed-autonity-contract
 
       - name: Build
         run: make build-docker-image

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,7 @@
 # Build Autonity in a stock Go builder container
 FROM golang:1.14-alpine as builder
 
-RUN apk add --no-cache make gcc musl-dev linux-headers libc-dev git
+RUN apk add --no-cache make gcc musl-dev linux-headers libc-dev git perl-utils
 
 ADD . /autonity
 RUN cd /autonity && make autonity

--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@
 .PHONY: autonity-windows autonity-windows-386 autonity-windows-amd64
 
 NPMBIN= $(shell npm bin)
-GOBIN = ./build/bin
+BINDIR = ./build/bin
 GO ?= latest
 LATEST_COMMIT ?= $(shell git log -n 1 develop --pretty=format:"%H")
 ifeq ($(LATEST_COMMIT),)
@@ -51,7 +51,7 @@ build-docker-image:
 autonity: embed-autonity-contract
 	build/env.sh go run build/ci.go install ./cmd/autonity
 	@echo "Done building."
-	@echo "Run \"$(GOBIN)/autonity\" to launch autonity."
+	@echo "Run \"$(BINDIR)/autonity\" to launch autonity."
 
 embed-autonity-contract: $(GENERATED_BYTECODE) $(GENERATED_ABI)
 
@@ -83,12 +83,12 @@ all: embed-autonity-contract
 android:
 	build/env.sh go run build/ci.go aar --local
 	@echo "Done building."
-	@echo "Import \"$(GOBIN)/autonity.aar\" to use the library."
+	@echo "Import \"$(BINDIR)/autonity.aar\" to use the library."
 
 ios:
 	build/env.sh go run build/ci.go xcode --local
 	@echo "Done building."
-	@echo "Import \"$(GOBIN)/autonity.framework\" to use the library."
+	@echo "Import \"$(BINDIR)/autonity.framework\" to use the library."
 
 test: all
 	build/env.sh go run build/ci.go test -coverage
@@ -161,19 +161,19 @@ lint-deps:
 
 clean:
 	go clean -cache
-	rm -fr build/_workspace/pkg/ $(GOBIN)/*
+	rm -fr build/_workspace/pkg/ $(BINDIR)/*
 	rm -rf $(GENERATED_CONTRACT_DIR)
 
 # The devtools target installs tools required for 'go generate'.
-# You need to put $GOBIN (or $GOPATH/bin) in your PATH to use 'go generate'.
+# You need to put $BINDIR (or $GOPATH/bin) in your PATH to use 'go generate'.
 
 devtools:
 	go get -u github.com/golang/mock/mockgen
-	env GOBIN= go get -u golang.org/x/tools/cmd/stringer
-	env GOBIN= go get -u github.com/kevinburke/go-bindata/go-bindata
-	env GOBIN= go get -u github.com/fjl/gencodec
-	env GOBIN= go get -u github.com/golang/protobuf/protoc-gen-go
-	env GOBIN= go install ./cmd/abigen
+	env BINDIR= go get -u golang.org/x/tools/cmd/stringer
+	env BINDIR= go get -u github.com/kevinburke/go-bindata/go-bindata
+	env BINDIR= go get -u github.com/fjl/gencodec
+	env BINDIR= go get -u github.com/golang/protobuf/protoc-gen-go
+	env BINDIR= go install ./cmd/abigen
 	@type "npm" 2> /dev/null || echo 'Please install node.js and npm'
 	@type "solc" 2> /dev/null || echo 'Please install solc'
 	@type "protoc" 2> /dev/null || echo 'Please install protoc'
@@ -182,90 +182,90 @@ devtools:
 
 autonity-cross: autonity-linux autonity-darwin autonity-windows autonity-android autonity-ios
 	@echo "Full cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-*
+	@ls -ld $(BINDIR)/autonity-*
 
 autonity-linux: autonity-linux-386 autonity-linux-amd64 autonity-linux-arm autonity-linux-mips64 autonity-linux-mips64le
 	@echo "Linux cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-*
+	@ls -ld $(BINDIR)/autonity-linux-*
 
 autonity-linux-386:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/386 -v ./cmd/autonity
 	@echo "Linux 386 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep 386
+	@ls -ld $(BINDIR)/autonity-linux-* | grep 386
 
 autonity-linux-amd64:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/amd64 -v ./cmd/autonity
 	@echo "Linux amd64 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep amd64
+	@ls -ld $(BINDIR)/autonity-linux-* | grep amd64
 
 autonity-linux-arm: autonity-linux-arm-5 autonity-linux-arm-6 autonity-linux-arm-7 autonity-linux-arm64
 	@echo "Linux ARM cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep arm
+	@ls -ld $(BINDIR)/autonity-linux-* | grep arm
 
 autonity-linux-arm-5:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/arm-5 -v ./cmd/autonity
 	@echo "Linux ARMv5 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep arm-5
+	@ls -ld $(BINDIR)/autonity-linux-* | grep arm-5
 
 autonity-linux-arm-6:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/arm-6 -v ./cmd/autonity
 	@echo "Linux ARMv6 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep arm-6
+	@ls -ld $(BINDIR)/autonity-linux-* | grep arm-6
 
 autonity-linux-arm-7:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/arm-7 -v ./cmd/autonity
 	@echo "Linux ARMv7 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep arm-7
+	@ls -ld $(BINDIR)/autonity-linux-* | grep arm-7
 
 autonity-linux-arm64:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/arm64 -v ./cmd/autonity
 	@echo "Linux ARM64 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep arm64
+	@ls -ld $(BINDIR)/autonity-linux-* | grep arm64
 
 autonity-linux-mips:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/mips --ldflags '-extldflags "-static"' -v ./cmd/autonity
 	@echo "Linux MIPS cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep mips
+	@ls -ld $(BINDIR)/autonity-linux-* | grep mips
 
 autonity-linux-mipsle:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/mipsle --ldflags '-extldflags "-static"' -v ./cmd/autonity
 	@echo "Linux MIPSle cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep mipsle
+	@ls -ld $(BINDIR)/autonity-linux-* | grep mipsle
 
 autonity-linux-mips64:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/mips64 --ldflags '-extldflags "-static"' -v ./cmd/autonity
 	@echo "Linux MIPS64 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep mips64
+	@ls -ld $(BINDIR)/autonity-linux-* | grep mips64
 
 autonity-linux-mips64le:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=linux/mips64le --ldflags '-extldflags "-static"' -v ./cmd/autonity
 	@echo "Linux MIPS64le cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-linux-* | grep mips64le
+	@ls -ld $(BINDIR)/autonity-linux-* | grep mips64le
 
 autonity-darwin: autonity-darwin-386 autonity-darwin-amd64
 	@echo "Darwin cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-darwin-*
+	@ls -ld $(BINDIR)/autonity-darwin-*
 
 autonity-darwin-386:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=darwin/386 -v ./cmd/autonity
 	@echo "Darwin 386 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-darwin-* | grep 386
+	@ls -ld $(BINDIR)/autonity-darwin-* | grep 386
 
 autonity-darwin-amd64:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=darwin/amd64 -v ./cmd/autonity
 	@echo "Darwin amd64 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-darwin-* | grep amd64
+	@ls -ld $(BINDIR)/autonity-darwin-* | grep amd64
 
 autonity-windows: autonity-windows-386 autonity-windows-amd64
 	@echo "Windows cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-windows-*
+	@ls -ld $(BINDIR)/autonity-windows-*
 
 autonity-windows-386:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=windows/386 -v ./cmd/autonity
 	@echo "Windows 386 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-windows-* | grep 386
+	@ls -ld $(BINDIR)/autonity-windows-* | grep 386
 
 autonity-windows-amd64:
 	build/env.sh go run build/ci.go xgo -- --go=$(GO) --targets=windows/amd64 -v ./cmd/autonity
 	@echo "Windows amd64 cross compilation done:"
-	@ls -ld $(GOBIN)/autonity-windows-* | grep amd64
+	@ls -ld $(BINDIR)/autonity-windows-* | grep amd64

--- a/Makefile
+++ b/Makefile
@@ -64,16 +64,18 @@ $(GENERATED_BYTECODE) $(GENERATED_ABI): $(AUTONITY_CONTRACT_DIR)/$(AUTONITY_CONT
 	$(DOCKER_SUDO) docker run --rm -v $(CURDIR)/$(AUTONITY_CONTRACT_DIR):/contracts -v $(CURDIR)/$(GENERATED_CONTRACT_DIR):/output ethereum/solc:0.6.4 --overwrite --abi --bin -o /output /contracts/$(AUTONITY_CONTRACT)
 
 	@echo Generating $(GENERATED_BYTECODE)
-	@echo 'package generated\n' > $(GENERATED_BYTECODE)
+	@echo 'package generated' > $(GENERATED_BYTECODE)
 	@echo -n 'const Bytecode = "' >> $(GENERATED_BYTECODE)
 	@cat  $(GENERATED_CONTRACT_DIR)/Autonity.bin >> $(GENERATED_BYTECODE)
 	@echo '"' >> $(GENERATED_BYTECODE)
+	@gofmt -s -w $(GENERATED_BYTECODE)
 
 	@echo Generating $(GENERATED_ABI)
-	@echo 'package generated\n' > $(GENERATED_ABI)
+	@echo 'package generated' > $(GENERATED_ABI)
 	@echo -n 'const Abi = `' >> $(GENERATED_ABI)
 	@cat  $(GENERATED_CONTRACT_DIR)/Autonity.abi | json_pp  >> $(GENERATED_ABI)
 	@echo '`' >> $(GENERATED_ABI)
+	@gofmt -s -w $(GENERATED_ABI)
 
 all: embed-autonity-contract
 	build/env.sh go run build/ci.go install

--- a/Makefile
+++ b/Makefile
@@ -37,6 +37,11 @@ GENERATED_BYTECODE = $(GENERATED_CONTRACT_DIR)/bytecode.go
 # we then echo "sudo".
 DOCKER_SUDO = $(shell [ `id -u` -eq 0 ] || id -nG $(USER) | grep "\<docker\>" > /dev/null || echo sudo )
 
+# Builds the docker image and checks that we can run the autonity binary inside it
+build-docker-image:
+	@$(DOCKER_SUDO) docker build -t autonity .
+	@$(DOCKER_SUDO) docker run --rm autonity -h > /dev/null
+
 autonity: embed-autonity-contract
 	build/env.sh go run build/ci.go install ./cmd/autonity
 	@echo "Done building."

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-# This Makefile is meant to be used by people that do not usually work
-# with Go source code. If you know what GOPATH is then you probably
-# don't need to bother with make.
+# Note: This makefile should refrain from using docker in any build step
+# required by the 'autonity' rule. This is so that it remains possible to run
+# 'make autonity' inside a docker build.
 
 .PHONY: autonity embed-autonity-contract android ios autonity-cross evm all test clean lint lint-deps mock-gen test-fast
 .PHONY: autonity-linux autonity-linux-386 autonity-linux-amd64 autonity-linux-mips64 autonity-linux-mips64le
@@ -15,6 +15,8 @@ LATEST_COMMIT ?= $(shell git log -n 1 develop --pretty=format:"%H")
 ifeq ($(LATEST_COMMIT),)
 LATEST_COMMIT := $(shell git log -n 1 HEAD~1 --pretty=format:"%H")
 endif
+SOLC_VERSION = 0.6.4
+SOLC_BINARY = $(BINDIR)/solc_static_linux_v$(SOLC_VERSION)
 
 AUTONITY_CONTRACT_DIR = ./contracts/autonity/contract/contracts
 AUTONITY_CONTRACT = Autonity.sol
@@ -38,13 +40,8 @@ GENERATED_BYTECODE = $(GENERATED_CONTRACT_DIR)/bytecode.go
 DOCKER_SUDO = $(shell [ `id -u` -eq 0 ] || id -nG $(USER) | grep "\<docker\>" > /dev/null || echo sudo )
 
 # Builds the docker image and checks that we can run the autonity binary inside
-# it. We need to run embed-autonity-contract before running the docker build
-# here since embed-autonity-contract makes use of "docker run" which cannot be
-# run inside of a "docker build", by running it before we ensure that the make
-# targets are up to date and that when running make inside the docker image
-# there is nothing to do for the embed-autonity-contract rule. 
+# it.
 build-docker-image:
-	make embed-autonity-contract
 	@$(DOCKER_SUDO) docker build -t autonity .
 	@$(DOCKER_SUDO) docker run --rm autonity -h > /dev/null
 
@@ -53,15 +50,12 @@ autonity: embed-autonity-contract
 	@echo "Done building."
 	@echo "Run \"$(BINDIR)/autonity\" to launch autonity."
 
+# Genreates go source files containing the contract bytecode and abi.
 embed-autonity-contract: $(GENERATED_BYTECODE) $(GENERATED_ABI)
 
-# NOTE previously we were using
-# https://github.com/ethereum/solidity/releases/download/v0.5.1/solc-static-linux
-# this binary does not produce the same bytecode as the ethereum/solc:0.5.1
-# docker image. This was causing tests to fail.
-$(GENERATED_BYTECODE) $(GENERATED_ABI): $(AUTONITY_CONTRACT_DIR)/$(AUTONITY_CONTRACT)
-	mkdir -p $(GENERATED_CONTRACT_DIR)
-	$(DOCKER_SUDO) docker run --rm -v $(CURDIR)/$(AUTONITY_CONTRACT_DIR):/contracts -v $(CURDIR)/$(GENERATED_CONTRACT_DIR):/output ethereum/solc:0.6.4 --overwrite --abi --bin -o /output /contracts/$(AUTONITY_CONTRACT)
+$(GENERATED_BYTECODE) $(GENERATED_ABI): $(AUTONITY_CONTRACT_DIR)/$(AUTONITY_CONTRACT) $(SOLC_BINARY)
+	@mkdir -p $(GENERATED_CONTRACT_DIR)
+	$(SOLC_BINARY) --overwrite --abi --bin -o $(GENERATED_CONTRACT_DIR) $(AUTONITY_CONTRACT_DIR)/$(AUTONITY_CONTRACT)
 
 	@echo Generating $(GENERATED_BYTECODE)
 	@echo 'package generated' > $(GENERATED_BYTECODE)
@@ -76,6 +70,11 @@ $(GENERATED_BYTECODE) $(GENERATED_ABI): $(AUTONITY_CONTRACT_DIR)/$(AUTONITY_CONT
 	@cat  $(GENERATED_CONTRACT_DIR)/Autonity.abi | json_pp  >> $(GENERATED_ABI)
 	@echo '`' >> $(GENERATED_ABI)
 	@gofmt -s -w $(GENERATED_ABI)
+
+$(SOLC_BINARY):
+	mkdir -p $(BINDIR)
+	wget -O $(SOLC_BINARY) https://github.com/ethereum/solidity/releases/download/v$(SOLC_VERSION)/solc-static-linux
+	chmod +x $(SOLC_BINARY)
 
 all: embed-autonity-contract
 	build/env.sh go run build/ci.go install

--- a/Makefile
+++ b/Makefile
@@ -22,14 +22,20 @@ GENERATED_CONTRACT_DIR = ./common/acdefault/generated
 GENERATED_ABI = $(GENERATED_CONTRACT_DIR)/abi.go
 GENERATED_BYTECODE = $(GENERATED_CONTRACT_DIR)/bytecode.go
 
-# This uses 'id -nG $USER' to list all groups that the user is part of and then
-# greps for the word docker in the output, if grep matches the word it returns
-# the successful error code and hence the command after the or '||' will not be
-# executed. This results in DOCKER_SUDO having the value 'sudo' if the current
-# user is not in the docker group and being empty if the current user is in the
-# docker group. This allows us to selectively execute docker with sudo for
-# users that need to execute it with sudo.
-DOCKER_SUDO = $(shell id -nG $(USER) | grep "\<docker\>" > /dev/null || echo sudo)
+# DOCKER_SUDO is set to either the empty string or "sudo" and is used to
+# control whether docker is executed with sudo or not. If the user is root or
+# the user is in the docker group then this will be set to the empty string,
+# otherwise it will be set to "sudo".
+#
+# We make use of posix's short-circuit evaluation of "or" expressions
+# (https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_09_03)
+# where the second part of the expression is only executed if the first part
+# evaluates to false. We first check to see if the user id is 0 since this
+# indicates that the user is root. If not we then use 'id -nG $USER' to list
+# all groups that the user is part of and then grep for the word docker in the
+# output, if grep matches the word it returns the successful error code. If not
+# we then echo "sudo".
+DOCKER_SUDO = $(shell [ `id -u` -eq 0 ] || id -nG $(USER) | grep "\<docker\>" > /dev/null || echo sudo )
 
 autonity: embed-autonity-contract
 	build/env.sh go run build/ci.go install ./cmd/autonity

--- a/Makefile
+++ b/Makefile
@@ -37,8 +37,14 @@ GENERATED_BYTECODE = $(GENERATED_CONTRACT_DIR)/bytecode.go
 # we then echo "sudo".
 DOCKER_SUDO = $(shell [ `id -u` -eq 0 ] || id -nG $(USER) | grep "\<docker\>" > /dev/null || echo sudo )
 
-# Builds the docker image and checks that we can run the autonity binary inside it
+# Builds the docker image and checks that we can run the autonity binary inside
+# it. We need to run embed-autonity-contract before running the docker build
+# here since embed-autonity-contract makes use of "docker run" which cannot be
+# run inside of a "docker build", by running it before we ensure that the make
+# targets are up to date and that when running make inside the docker image
+# there is nothing to do for the embed-autonity-contract rule. 
 build-docker-image:
+	make embed-autonity-contract
 	@$(DOCKER_SUDO) docker build -t autonity .
 	@$(DOCKER_SUDO) docker run --rm autonity -h > /dev/null
 

--- a/README.md
+++ b/README.md
@@ -30,6 +30,12 @@ This generates go source from the autonity contract.
 make autonity
 ```
 
+## Build Autonity docker image
+
+```
+make build-docker-image
+```
+
 ## License
 
 The go-ethereum library (i.e. all code outside of the `cmd` directory) is licensed under the

